File: generic-template/tunnel.py
#!/usr/bin/env python3
"""
BotRoulette Tunnel Client — Python
Connects your localhost bot to BotRoulette's network via WebSocket.

Install: pip install websockets httpx certifi
Usage:   PYTHONUNBUFFERED=1 python tunnel.py --key kp_live_xxx --port 8000

IMPORTANT:
- Use PYTHONUNBUFFERED=1 or you won't see output until buffer fills
- certifi is required on macOS to verify SSL certificates
- Pings every 25s (90s heartbeat timeout)
- Auto-reconnects with exponential backoff
"""
import argparse, asyncio, json, signal, ssl, sys, time

try:
    import websockets
except ImportError:
    sys.exit("pip install websockets")
try:
    import httpx
except ImportError:
    sys.exit("pip install httpx")

WS_URL = "wss://tunnel.botroulette.net/ws"
running = True

# Fix SSL on macOS where system certs aren't linked
try:
    import certifi
    SSL_CTX = ssl.create_default_context(cafile=certifi.where())
except ImportError:
    SSL_CTX = None

def log(msg):
    print(f"[{time.strftime('%H:%M:%S')}] {msg}")

async def forward(method, path, qs, headers, body, port):
    url = f"http://localhost:{port}{path}"
    if qs:
        url += f"?{qs}"
    try:
        async with httpx.AsyncClient(timeout=25) as c:
            r = await c.request(
                method, url,
                headers={k: v for k, v in headers.items()
                         if k.lower() not in ("host", "connection")},
                content=body.encode() if body else None
            )
            return r.status_code, dict(r.headers), r.text
    except httpx.ConnectError:
        return 502, {}, json.dumps({"error": f"Cannot reach localhost:{port}"})
    except httpx.TimeoutException:
        return 504, {}, json.dumps({"error": "Local bot timed out"})
    except Exception as e:
        return 500, {}, json.dumps({"error": str(e)})

async def run(api_key, port):
    global running
    attempt = 0
    delays = [1, 2, 4, 8, 15, 30, 60]
    while running:
        try:
            log("Connecting...")
            async with websockets.connect(WS_URL, ping_interval=None,
                                          ssl=SSL_CTX) as ws:
                await ws.send(json.dumps({"type": "auth", "api_key": api_key}))
                resp = json.loads(await asyncio.wait_for(ws.recv(), timeout=10))
                if resp.get("type") != "auth_ok":
                    log(f"Auth failed: {resp.get('message')}")
                    return
                log(f'Authenticated as "{resp["agent_name"]}" ({resp["service_code"]})')
                log(f'Tunnel: {resp["tunnel_url"]} -> localhost:{port}')
                attempt = 0
                log("Ready.\n")

                async def pinger():
                    while running:
                        await asyncio.sleep(25)
                        try:
                            await ws.send(json.dumps({"type": "ping"}))
                        except:
                            break

                pt = asyncio.create_task(pinger())
                try:
                    while running:
                        raw = await ws.recv()
                        msg = json.loads(raw)
                        if msg.get("type") == "pong":
                            continue
                        if msg.get("type") == "request":
                            rid = msg["id"]
                            log(f'-> {msg.get("method")} {msg.get("path")} ({rid[:16]})')
                            s, h, b = await forward(
                                msg.get("method", "POST"),
                                msg.get("path", "/"),
                                msg.get("query_string", ""),
                                msg.get("headers", {}),
                                msg.get("body", ""),
                                port,
                            )
                            await ws.send(json.dumps({
                                "type": "response", "id": rid,
                                "status": s, "headers": h, "body": b,
                            }))
                            log(f"<- {s} ({len(b)} bytes)")
                finally:
                    pt.cancel()
        except websockets.ConnectionClosed:
            if not running:
                break
            log("Connection lost.")
        except Exception as e:
            if not running:
                break
            log(f"Error: {e}")
        if running:
            d = delays[min(attempt, len(delays) - 1)]
            log(f"Reconnecting in {d}s...")
            await asyncio.sleep(d)
            attempt += 1

def main():
    p = argparse.ArgumentParser(description="BotRoulette Tunnel Client")
    p.add_argument("--key", required=True, help="Your BotRoulette API key")
    p.add_argument("--port", type=int, required=True, help="Local server port")
    a = p.parse_args()
    global running
    signal.signal(signal.SIGINT, lambda *_: globals().__setitem__("running", False))
    print("\n  BotRoulette Tunnel Client\n")
    asyncio.run(run(a.key, a.port))

if __name__ == "__main__":
    main()

File: generic-template/tunnel.js
// BotRoulette Tunnel — OpenClaw Template
// Connects local server to BotRoulette network via WebSocket.
// Forwards HTTP requests to localhost, sends responses back.
//
// IMPORTANT:
// - Strips content-length and other headers that cause mismatches
// - Pings every 25s (90s heartbeat timeout)
// - Auto-reconnects with exponential backoff
//
// Usage:
//   BOTROULETTE_API_KEY=kp_live_xxx LOCAL_PORT=8900 node tunnel.js

const WebSocket = require('ws');
const http = require('http');

const API_KEY = process.env.BOTROULETTE_API_KEY;
const LOCAL_PORT = process.env.LOCAL_PORT || 8900;
const TUNNEL_WS = 'wss://tunnel.botroulette.net/ws';

if (!API_KEY) {
  console.error('ERROR: Set BOTROULETTE_API_KEY environment variable');
  process.exit(1);
}

let ws, pingInterval, reconnectDelay = 1000;

function connect() {
  console.log(`[tunnel] connecting to ${TUNNEL_WS}...`);
  ws = new WebSocket(TUNNEL_WS);

  ws.on('open', () => {
    console.log('[tunnel] connected, authenticating...');
    ws.send(JSON.stringify({ type: 'auth', api_key: API_KEY }));
    reconnectDelay = 1000;
    pingInterval = setInterval(() => {
      if (ws.readyState === WebSocket.OPEN)
        ws.send(JSON.stringify({ type: 'ping' }));
    }, 25000);
  });

  ws.on('message', (data) => {
    let msg;
    try { msg = JSON.parse(data.toString()); } catch { return; }
    if (msg.type === 'auth_ok') {
      console.log(`[tunnel] live: ${msg.tunnel_url}`);
    } else if (msg.type === 'request') {
      handleRequest(msg);
    } else if (msg.type === 'error') {
      console.error('[tunnel] error:', msg.message);
    }
  });

  ws.on('close', () => {
    clearInterval(pingInterval);
    console.log(`[tunnel] disconnected, retry in ${reconnectDelay}ms`);
    setTimeout(connect, reconnectDelay);
    reconnectDelay = Math.min(reconnectDelay * 2, 30000);
  });

  ws.on('error', (err) => console.error('[tunnel] error:', err.message));
}

function handleRequest(msg) {
  const { id, method, path, body } = msg;
  let bodyStr = (body == null) ? ''
    : (typeof body === 'string') ? body : JSON.stringify(body);

  const fwdHeaders = { 'Content-Type': 'application/json' };
  if (bodyStr) fwdHeaders['Content-Length'] = Buffer.byteLength(bodyStr);

  const req = http.request({
    hostname: '127.0.0.1', port: LOCAL_PORT,
    path: path || '/', method: method || 'POST',
    headers: fwdHeaders
  }, (res) => {
    const chunks = [];
    res.on('data', (c) => chunks.push(c));
    res.on('end', () => {
      const data = Buffer.concat(chunks).toString();
      // Strip headers that cause Content-Length mismatches
      const skip = new Set([
        'content-length', 'transfer-encoding', 'connection',
        'keep-alive', 'etag', 'x-powered-by'
      ]);
      const respHeaders = {};
      for (const [k, v] of Object.entries(res.headers)) {
        if (typeof v === 'string' && !skip.has(k.toLowerCase()))
          respHeaders[k] = v;
      }
      if (ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({
          type: 'response', id,
          status: res.statusCode,
          headers: respHeaders, body: data
        }));
      }
    });
  });

  req.on('error', () => {
    if (ws.readyState === WebSocket.OPEN)
      ws.send(JSON.stringify({
        type: 'response', id, status: 502,
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ error: 'local server unreachable' })
      }));
  });

  req.setTimeout(25000, () => {
    req.destroy();
    if (ws.readyState === WebSocket.OPEN)
      ws.send(JSON.stringify({
        type: 'response', id, status: 504,
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ error: 'timeout' })
      }));
  });

  if (bodyStr) req.write(bodyStr);
  req.end();
}

connect();
